File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-variance-swaps-5-4/VarianceSwapTransactionSupplement.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class VarianceSwapTransactionSupplement : Product
    {
        public VarianceSwapTransactionSupplement(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNodeList varianceLegNodeList = xmlNode.SelectNodes("varianceLeg");
            
            if (varianceLegNodeList != null)
            {
                this.varianceLeg_ = new List<VarianceLeg>();
                foreach (XmlNode item in varianceLegNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            varianceLegIDRef_ = item.Attributes["id"].Value;
                            varianceLeg_.Add(new VarianceLeg(item));
                            IDManager.SetID(varianceLegIDRef_, varianceLeg_[varianceLeg_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            varianceLegIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        varianceLeg_.Add(new VarianceLeg(item));
                        }
                    }
                    else
                    {
                        varianceLeg_.Add(new VarianceLeg(item));
                    }
                }
            }
            
        
            XmlNode multipleExchangeIndexAnnexFallbackNode = xmlNode.SelectSingleNode("multipleExchangeIndexAnnexFallback");
            
            if (multipleExchangeIndexAnnexFallbackNode != null)
            {
                if (multipleExchangeIndexAnnexFallbackNode.Attributes["href"] != null || multipleExchangeIndexAnnexFallbackNode.Attributes["id"] != null) 
                {
                    if (multipleExchangeIndexAnnexFallbackNode.Attributes["id"] != null) 
                    {
                        multipleExchangeIndexAnnexFallbackIDRef_ = multipleExchangeIndexAnnexFallbackNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(multipleExchangeIndexAnnexFallbackNode);
                        IDManager.SetID(multipleExchangeIndexAnnexFallbackIDRef_, ob);
                    }
                    else if (multipleExchangeIndexAnnexFallbackNode.Attributes["href"] != null)
                    {
                        multipleExchangeIndexAnnexFallbackIDRef_ = multipleExchangeIndexAnnexFallbackNode.Attributes["href"].Value;
                    }
                    else
                    {
                        multipleExchangeIndexAnnexFallback_ = new XsdTypeBoolean(multipleExchangeIndexAnnexFallbackNode);
                    }
                }
                else
                {
                    multipleExchangeIndexAnnexFallback_ = new XsdTypeBoolean(multipleExchangeIndexAnnexFallbackNode);
                }
            }
            
        
            XmlNode componentSecurityIndexAnnexFallbackNode = xmlNode.SelectSingleNode("componentSecurityIndexAnnexFallback");
            
            if (componentSecurityIndexAnnexFallbackNode != null)
            {
                if (componentSecurityIndexAnnexFallbackNode.Attributes["href"] != null || componentSecurityIndexAnnexFallbackNode.Attributes["id"] != null) 
                {
                    if (componentSecurityIndexAnnexFallbackNode.Attributes["id"] != null) 
                    {
                        componentSecurityIndexAnnexFallbackIDRef_ = componentSecurityIndexAnnexFallbackNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(componentSecurityIndexAnnexFallbackNode);
                        IDManager.SetID(componentSecurityIndexAnnexFallbackIDRef_, ob);
                    }
                    else if (componentSecurityIndexAnnexFallbackNode.Attributes["href"] != null)
                    {
                        componentSecurityIndexAnnexFallbackIDRef_ = componentSecurityIndexAnnexFallbackNode.Attributes["href"].Value;
                    }
                    else
                    {
                        componentSecurityIndexAnnexFallback_ = new XsdTypeBoolean(componentSecurityIndexAnnexFallbackNode);
                    }
                }
                else
                {
                    componentSecurityIndexAnnexFallback_ = new XsdTypeBoolean(componentSecurityIndexAnnexFallbackNode);
                }
            }
            
        
            XmlNode localJurisdictionNode = xmlNode.SelectSingleNode("localJurisdiction");
            
            if (localJurisdictionNode != null)
            {
                if (localJurisdictionNode.Attributes["href"] != null || localJurisdictionNode.Attributes["id"] != null) 
                {
                    if (localJurisdictionNode.Attributes["id"] != null) 
                    {
                        localJurisdictionIDRef_ = localJurisdictionNode.Attributes["id"].Value;
                        CountryCode ob = new CountryCode(localJurisdictionNode);
                        IDManager.SetID(localJurisdictionIDRef_, ob);
                    }
                    else if (localJurisdictionNode.Attributes["href"] != null)
                    {
                        localJurisdictionIDRef_ = localJurisdictionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        localJurisdiction_ = new CountryCode(localJurisdictionNode);
                    }
                }
                else
                {
                    localJurisdiction_ = new CountryCode(localJurisdictionNode);
                }
            }
            
        
            XmlNode relevantJurisdictionNode = xmlNode.SelectSingleNode("relevantJurisdiction");
            
            if (relevantJurisdictionNode != null)
            {
                if (relevantJurisdictionNode.Attributes["href"] != null || relevantJurisdictionNode.Attributes["id"] != null) 
                {
                    if (relevantJurisdictionNode.Attributes["id"] != null) 
                    {
                        relevantJurisdictionIDRef_ = relevantJurisdictionNode.Attributes["id"].Value;
                        CountryCode ob = new CountryCode(relevantJurisdictionNode);
                        IDManager.SetID(relevantJurisdictionIDRef_, ob);
                    }
                    else if (relevantJurisdictionNode.Attributes["href"] != null)
                    {
                        relevantJurisdictionIDRef_ = relevantJurisdictionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        relevantJurisdiction_ = new CountryCode(relevantJurisdictionNode);
                    }
                }
                else
                {
                    relevantJurisdiction_ = new CountryCode(relevantJurisdictionNode);
                }
            }
            
        
        }
        
    
        #region VarianceLeg_
        private List<VarianceLeg> varianceLeg_;
        public List<VarianceLeg> VarianceLeg_
        {
            get
            {
                if (this.varianceLeg_ != null)
                {
                    return this.varianceLeg_; 
                }
                else if (this.varianceLegIDRef_ != null)
                {
                    return this.varianceLeg_; 
                }
                else
                {
                      return this.varianceLeg_; 
                }
            }
            set
            {
                if (this.varianceLeg_ != value)
                {
                    this.varianceLeg_ = value;
                }
            }
        }
        #endregion
        
        public string varianceLegIDRef_ { get; set; }
        #region MultipleExchangeIndexAnnexFallback_
        private XsdTypeBoolean multipleExchangeIndexAnnexFallback_;
        public XsdTypeBoolean MultipleExchangeIndexAnnexFallback_
        {
            get
            {
                if (this.multipleExchangeIndexAnnexFallback_ != null)
                {
                    return this.multipleExchangeIndexAnnexFallback_; 
                }
                else if (this.multipleExchangeIndexAnnexFallbackIDRef_ != null)
                {
                    multipleExchangeIndexAnnexFallback_ = IDManager.getID(multipleExchangeIndexAnnexFallbackIDRef_) as XsdTypeBoolean;
                    return this.multipleExchangeIndexAnnexFallback_; 
                }
                else
                {
                      return this.multipleExchangeIndexAnnexFallback_; 
                }
            }
            set
            {
                if (this.multipleExchangeIndexAnnexFallback_ != value)
                {
                    this.multipleExchangeIndexAnnexFallback_ = value;
                }
            }
        }
        #endregion
        
        public string multipleExchangeIndexAnnexFallbackIDRef_ { get; set; }
        #region ComponentSecurityIndexAnnexFallback_
        private XsdTypeBoolean componentSecurityIndexAnnexFallback_;
        public XsdTypeBoolean ComponentSecurityIndexAnnexFallback_
        {
            get
            {
                if (this.componentSecurityIndexAnnexFallback_ != null)
                {
                    return this.componentSecurityIndexAnnexFallback_; 
                }
                else if (this.componentSecurityIndexAnnexFallbackIDRef_ != null)
                {
                    componentSecurityIndexAnnexFallback_ = IDManager.getID(componentSecurityIndexAnnexFallbackIDRef_) as XsdTypeBoolean;
                    return this.componentSecurityIndexAnnexFallback_; 
                }
                else
                {
                      return this.componentSecurityIndexAnnexFallback_; 
                }
            }
            set
            {
                if (this.componentSecurityIndexAnnexFallback_ != value)
                {
                    this.componentSecurityIndexAnnexFallback_ = value;
                }
            }
        }
        #endregion
        
        public string componentSecurityIndexAnnexFallbackIDRef_ { get; set; }
        #region LocalJurisdiction_
        private CountryCode localJurisdiction_;
        public CountryCode LocalJurisdiction_
        {
            get
            {
                if (this.localJurisdiction_ != null)
                {
                    return this.localJurisdiction_; 
                }
                else if (this.localJurisdictionIDRef_ != null)
                {
                    localJurisdiction_ = IDManager.getID(localJurisdictionIDRef_) as CountryCode;
                    return this.localJurisdiction_; 
                }
                else
                {
                      return this.localJurisdiction_; 
                }
            }
            set
            {
                if (this.localJurisdiction_ != value)
                {
                    this.localJurisdiction_ = value;
                }
            }
        }
        #endregion
        
        public string localJurisdictionIDRef_ { get; set; }
        #region RelevantJurisdiction_
        private CountryCode relevantJurisdiction_;
        public CountryCode RelevantJurisdiction_
        {
            get
            {
                if (this.relevantJurisdiction_ != null)
                {
                    return this.relevantJurisdiction_; 
                }
                else if (this.relevantJurisdictionIDRef_ != null)
                {
                    relevantJurisdiction_ = IDManager.getID(relevantJurisdictionIDRef_) as CountryCode;
                    return this.relevantJurisdiction_; 
                }
                else
                {
                      return this.relevantJurisdiction_; 
                }
            }
            set
            {
                if (this.relevantJurisdiction_ != value)
                {
                    this.relevantJurisdiction_ = value;
                }
            }
        }
        #endregion
        
        public string relevantJurisdictionIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-variance-swaps-5-4/VarianceSwapTransactionSupplement.cpp
// VarianceSwapTransactionSupplement.cpp 
#include "VarianceSwapTransactionSupplement.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

VarianceSwapTransactionSupplement::VarianceSwapTransactionSupplement(TiXmlNode* xmlNode)
: Product(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //varianceLegNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* varianceLegNode = xmlNode->FirstChildElement("varianceLeg");

   if(varianceLegNode){varianceLegIsNull_ = false;}
   else{varianceLegIsNull_ = true;}

   if(varianceLegNode)
   {
      for(varianceLegNode; varianceLegNode; varianceLegNode = varianceLegNode->NextSiblingElement("varianceLeg")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- varianceLegNode , address : " << varianceLegNode << std::endl;
          #endif
          if(varianceLegNode->Attribute("href") || varianceLegNode->Attribute("id"))
          {
              if(varianceLegNode->Attribute("id"))
              {
                  varianceLegIDRef_ = varianceLegNode->Attribute("id");
                  varianceLeg_.push_back(boost::shared_ptr<VarianceLeg>(new VarianceLeg(varianceLegNode)));
                  varianceLeg_.back()->setID(varianceLegIDRef_);
                  IDManager::instance().setID(varianceLegIDRef_, varianceLeg_.back());
              }
              else if(varianceLegNode->Attribute("href")) { varianceLegIDRef_ = varianceLegNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { varianceLeg_.push_back(boost::shared_ptr<VarianceLeg>(new VarianceLeg(varianceLegNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- varianceLegNode , address : " << varianceLegNode << std::endl;
       #endif
   }

   //multipleExchangeIndexAnnexFallbackNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* multipleExchangeIndexAnnexFallbackNode = xmlNode->FirstChildElement("multipleExchangeIndexAnnexFallback");

   if(multipleExchangeIndexAnnexFallbackNode){multipleExchangeIndexAnnexFallbackIsNull_ = false;}
   else{multipleExchangeIndexAnnexFallbackIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- multipleExchangeIndexAnnexFallbackNode , address : " << multipleExchangeIndexAnnexFallbackNode << std::endl;
   #endif
   if(multipleExchangeIndexAnnexFallbackNode)
   {
      if(multipleExchangeIndexAnnexFallbackNode->Attribute("href") || multipleExchangeIndexAnnexFallbackNode->Attribute("id"))
      {
          if(multipleExchangeIndexAnnexFallbackNode->Attribute("id"))
          {
             multipleExchangeIndexAnnexFallbackIDRef_ = multipleExchangeIndexAnnexFallbackNode->Attribute("id");
             multipleExchangeIndexAnnexFallback_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(multipleExchangeIndexAnnexFallbackNode));
             multipleExchangeIndexAnnexFallback_->setID(multipleExchangeIndexAnnexFallbackIDRef_);
             IDManager::instance().setID(multipleExchangeIndexAnnexFallbackIDRef_,multipleExchangeIndexAnnexFallback_);
          }
          else if(multipleExchangeIndexAnnexFallbackNode->Attribute("href")) { multipleExchangeIndexAnnexFallbackIDRef_ = multipleExchangeIndexAnnexFallbackNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { multipleExchangeIndexAnnexFallback_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(multipleExchangeIndexAnnexFallbackNode));}
   }

   //componentSecurityIndexAnnexFallbackNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* componentSecurityIndexAnnexFallbackNode = xmlNode->FirstChildElement("componentSecurityIndexAnnexFallback");

   if(componentSecurityIndexAnnexFallbackNode){componentSecurityIndexAnnexFallbackIsNull_ = false;}
   else{componentSecurityIndexAnnexFallbackIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- componentSecurityIndexAnnexFallbackNode , address : " << componentSecurityIndexAnnexFallbackNode << std::endl;
   #endif
   if(componentSecurityIndexAnnexFallbackNode)
   {
      if(componentSecurityIndexAnnexFallbackNode->Attribute("href") || componentSecurityIndexAnnexFallbackNode->Attribute("id"))
      {
          if(componentSecurityIndexAnnexFallbackNode->Attribute("id"))
          {
             componentSecurityIndexAnnexFallbackIDRef_ = componentSecurityIndexAnnexFallbackNode->Attribute("id");
             componentSecurityIndexAnnexFallback_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(componentSecurityIndexAnnexFallbackNode));
             componentSecurityIndexAnnexFallback_->setID(componentSecurityIndexAnnexFallbackIDRef_);
             IDManager::instance().setID(componentSecurityIndexAnnexFallbackIDRef_,componentSecurityIndexAnnexFallback_);
          }
          else if(componentSecurityIndexAnnexFallbackNode->Attribute("href")) { componentSecurityIndexAnnexFallbackIDRef_ = componentSecurityIndexAnnexFallbackNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { componentSecurityIndexAnnexFallback_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(componentSecurityIndexAnnexFallbackNode));}
   }

   //localJurisdictionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* localJurisdictionNode = xmlNode->FirstChildElement("localJurisdiction");

   if(localJurisdictionNode){localJurisdictionIsNull_ = false;}
   else{localJurisdictionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- localJurisdictionNode , address : " << localJurisdictionNode << std::endl;
   #endif
   if(localJurisdictionNode)
   {
      if(localJurisdictionNode->Attribute("href") || localJurisdictionNode->Attribute("id"))
      {
          if(localJurisdictionNode->Attribute("id"))
          {
             localJurisdictionIDRef_ = localJurisdictionNode->Attribute("id");
             localJurisdiction_ = boost::shared_ptr<CountryCode>(new CountryCode(localJurisdictionNode));
             localJurisdiction_->setID(localJurisdictionIDRef_);
             IDManager::instance().setID(localJurisdictionIDRef_,localJurisdiction_);
          }
          else if(localJurisdictionNode->Attribute("href")) { localJurisdictionIDRef_ = localJurisdictionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { localJurisdiction_ = boost::shared_ptr<CountryCode>(new CountryCode(localJurisdictionNode));}
   }

   //relevantJurisdictionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* relevantJurisdictionNode = xmlNode->FirstChildElement("relevantJurisdiction");

   if(relevantJurisdictionNode){relevantJurisdictionIsNull_ = false;}
   else{relevantJurisdictionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- relevantJurisdictionNode , address : " << relevantJurisdictionNode << std::endl;
   #endif
   if(relevantJurisdictionNode)
   {
      if(relevantJurisdictionNode->Attribute("href") || relevantJurisdictionNode->Attribute("id"))
      {
          if(relevantJurisdictionNode->Attribute("id"))
          {
             relevantJurisdictionIDRef_ = relevantJurisdictionNode->Attribute("id");
             relevantJurisdiction_ = boost::shared_ptr<CountryCode>(new CountryCode(relevantJurisdictionNode));
             relevantJurisdiction_->setID(relevantJurisdictionIDRef_);
             IDManager::instance().setID(relevantJurisdictionIDRef_,relevantJurisdiction_);
          }
          else if(relevantJurisdictionNode->Attribute("href")) { relevantJurisdictionIDRef_ = relevantJurisdictionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { relevantJurisdiction_ = boost::shared_ptr<CountryCode>(new CountryCode(relevantJurisdictionNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
std::vector<boost::shared_ptr<VarianceLeg>> VarianceSwapTransactionSupplement::getVarianceLeg()
{
   if(!this->varianceLegIsNull_){
        if(varianceLegIDRef_ != ""){
             return this->varianceLeg_;
        }else{
             return this->varianceLeg_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<VarianceLeg>>();
   }
}
boost::shared_ptr<XsdTypeBoolean> VarianceSwapTransactionSupplement::getMultipleExchangeIndexAnnexFallback()
{
   if(!this->multipleExchangeIndexAnnexFallbackIsNull_){
        if(multipleExchangeIndexAnnexFallbackIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(multipleExchangeIndexAnnexFallbackIDRef_));
        }else{
             return this->multipleExchangeIndexAnnexFallback_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
boost::shared_ptr<XsdTypeBoolean> VarianceSwapTransactionSupplement::getComponentSecurityIndexAnnexFallback()
{
   if(!this->componentSecurityIndexAnnexFallbackIsNull_){
        if(componentSecurityIndexAnnexFallbackIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(componentSecurityIndexAnnexFallbackIDRef_));
        }else{
             return this->componentSecurityIndexAnnexFallback_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
boost::shared_ptr<CountryCode> VarianceSwapTransactionSupplement::getLocalJurisdiction()
{
   if(!this->localJurisdictionIsNull_){
        if(localJurisdictionIDRef_ != ""){
             return boost::shared_static_cast<CountryCode>(IDManager::instance().getID(localJurisdictionIDRef_));
        }else{
             return this->localJurisdiction_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CountryCode>();
   }
}
boost::shared_ptr<CountryCode> VarianceSwapTransactionSupplement::getRelevantJurisdiction()
{
   if(!this->relevantJurisdictionIsNull_){
        if(relevantJurisdictionIDRef_ != ""){
             return boost::shared_static_cast<CountryCode>(IDManager::instance().getID(relevantJurisdictionIDRef_));
        }else{
             return this->relevantJurisdiction_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CountryCode>();
   }
}
}

